File: runtimes/java-11.0/src/main/java/io/openruntimes/java/RuntimeResponse.java
package io.openruntimes.java;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.HashMap;
import java.util.Map;

public class RuntimeResponse {
    private static final Gson gson = new GsonBuilder().serializeNulls().create();

    public RuntimeOutput send(String body, int statusCode, Map<String, String> headers) {
        return new RuntimeOutput(body, statusCode, headers);
    }

    public RuntimeOutput send(String body, int statusCode) {
        return this.send(body, statusCode, new HashMap<>());
    }

    public RuntimeOutput send(String body) {
        return this.send(body, 200, new HashMap<>());
    }

    public RuntimeOutput json(Map<String, Object> json, int statusCode, Map<String, String> headers) {
        if (!headers.containsKey("content-type")) {
            headers.put("content-type", "application/json");
        }
        return this.send(gson.toJson(json), statusCode, headers);
    }

    public RuntimeOutput json(Map<String, Object> json, int statusCode) {
        return this.json(json, statusCode, new HashMap<>());
    }

    public RuntimeOutput json(Map<String, Object> json) {
        return this.json(json, 200, new HashMap<>());
    }

    public RuntimeOutput empty() {
        return this.send("", 204, new HashMap<>());
    }

    public RuntimeOutput redirect(String url, int statusCode, Map<String, String> headers) {
        headers.put("location", url);
        return this.send("", statusCode, headers);
    }

    public RuntimeOutput redirect(String url, int statusCode) {
        return this.redirect(url, statusCode, new HashMap<>());
    }

    public RuntimeOutput redirect(String url) {
        return this.redirect(url, 301, new HashMap<>());
    }
}

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/RuntimeResponse.kt
package io.openruntimes.kotlin

import com.google.gson.GsonBuilder

public class RuntimeResponse {
    companion object {
        private val gson = GsonBuilder().serializeNulls().create()
    }

    fun binary(
        body: ByteArray,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf()
    ): RuntimeOutput {
        return RuntimeOutput(body, statusCode, headers)
    }

    fun text(
        body: String,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf()
    ): RuntimeOutput {
        return binary(body.toByteArray(), statusCode, headers)
    }

    fun send(
        body: String,
        statusCode: Int = 200,
        headers: Map<String, String> = mapOf()
    ): RuntimeOutput {
        return text(body, statusCode, headers)
    }

    fun json(
        json: MutableMap<String, Any>,
        statusCode: Int = 200,
        headers: MutableMap<String, String> = mutableMapOf()
    ): RuntimeOutput {
        headers["content-type"] = "application/json"
        return this.text(gson.toJson(json), statusCode, headers)
    }

    fun empty(): RuntimeOutput {
        return this.text("", 204, mutableMapOf<String, String>())
    }

    fun redirect(
        url: String,
        statusCode: Int = 301,
        headers: MutableMap<String, String> = mutableMapOf()
    ): RuntimeOutput {
        headers["location"] = url
        return this.text("", statusCode, headers)
    }
}